src/ray/raylet/node_manager.cc

```diff
@@ -1421,9 +1424,31 @@ void NodeManager::TreatTaskAsFailed(const Task &task, const ErrorType &error_typ
     // information about the TaskSpecification implementation.
     num_returns -= 1;
   }
+  // Determine which IDs should be marked as failed.
+  std::vector<plasma::ObjectID> objects_to_fail;
+  if (num_returns == 0 && required_object_id != nullptr) {
+    objects_to_fail.push_back(required_object_id->ToPlasmaId());
+  } else {
+    for (int64_t i = 0; i < num_returns; i++) {
+      objects_to_fail.push_back(spec.ReturnId(i).ToPlasmaId());
+    }
+  }
+  const JobID job_id = task.GetTaskSpecification().JobId();
+  MarkObjectsAsFailed(error_type, objects_to_fail, &job_id);
+  task_dependency_manager_.TaskCanceled(spec.TaskId());
+  // Notify the task dependency manager that we no longer need this task's
+  // object dependencies. TODO(swang): Ideally, we would check the return value
+  // here. However, we don't know at this point if the task was in the WAITING
+  // or READY queue before, in which case we would not have been subscribed to
+  // its dependencies.
+  task_dependency_manager_.UnsubscribeDependencies(spec.TaskId());
+}
+
+void NodeManager::MarkObjectsAsFailed(const ErrorType &error_type,
+                                      const std::vector<plasma::ObjectID> objects_to_fail,
+                                      const JobID *job_id) {
```

Comment:
I think this would be nicer if we passed in `const JobID &job_id` and used `JobID::Nil()` as the null value. Also, it probably makes sense to still push the error to the GCS even if there's no matching driver.